### PR TITLE
optimize druid query for complex measure

### PR DESCRIPTION
**Pull request by pjain1**
Fixes https://github.com/rilldata/rill-private-issues/issues/732

### CODE REVIEW

**Review by begelundmuller — APPROVED** [review]
LGTM

File: runtime/metricsview/executor_rewrite_druid_groups.go [hunk 1]
```diff
@@ -30,12 +31,6 @@ func (e *Executor) rewriteDruidGroupsWalk(n *SelectNode) error {
 	}
 
 	// Recurse
-	if n.FromSelect != nil {
-		err := e.rewriteDruidGroupsWalk(n.FromSelect)
-		if err != nil {
-			return err
-		}
-	}
```
Comment (begelundmuller) [inline runtime/metricsview/executor_rewrite_druid_groups.go:38]: Or wait, shouldn't it recurse here? The `FromSelect` might have nested joins

File: runtime/metricsview/executor_rewrite_druid_groups.go [hunk 2]
```diff
@@ -17,8 +17,9 @@ func (e *Executor) rewriteDruidGroups(ast *AST) error {
 }
 
 func (e *Executor) rewriteDruidGroupsWalk(n *SelectNode) error {
-	// Skip if already grouped
-	if n.Group {
+	// Skip if already grouped or if there is no sub query with JOIN
+	// Druid requires GROUP BY and ANY_VALUE aggregate for measure if there is a subquery with JOIN
+	if n.Group || (n.SpineSelect == nil && len(n.LeftJoinSelects) == 0 && n.JoinComparisonSelect == nil) {
```
Comment (begelundmuller) [inline runtime/metricsview/executor_rewrite_druid_groups.go:22]: And what about `SpineSelect`? That is `RIGHT JOIN`ed to the query

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: runtime/metricsview/executor_rewrite_druid_groups.go (hunk 2 above)
Comment (pjain1) [reply runtime/metricsview/executor_rewrite_druid_groups.go]: Just checked with measure filter, it seems like we may not add group by here however the query is getting converted to a new `windowOperator` query and I am already hitting some edge cases with this query. We may need to look into upgrading druid cluster to get latest version of this query, till then lets not make change for this.

File: runtime/metricsview/executor_rewrite_druid_groups.go [hunk 3]
```diff
@@ -22,6 +22,20 @@ func (e *Executor) rewriteDruidGroupsWalk(n *SelectNode) error {
 		return nil
 	}
 
+	// Recurse over FromSelect as it may contain sub queries with joins
+	if n.FromSelect != nil {
+		err := e.rewriteDruidGroupsWalk(n.FromSelect)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (begelundmuller) [inline runtime/metricsview/executor_rewrite_druid_groups.go:31]: Now there are two "// Recurse ..." sections in this function – what about just moving the `// Recurse` section above the `// Rewrite` section, and having a `var hasJoins bool` that gets set in the join recurse cases?

**Review by begelundmuller — APPROVED** [review]

**Pull request merged**